site/docs/plugins/conversations.md

```diff
@@ -0,0 +1,900 @@
+  return message?.text === "42";
+}
+```
+
+</CodeGroupItem>
+</CodeGroup>
+
+It returns `true` if the user may pass, and `false` otherwise.
+You can now use it in your main conversation builder function like this:
+
+<CodeGroup>
+  <CodeGroupItem title="TypeScript" active>
+
+```ts
+async function enterGroup(conversation: MyConversation, ctx: MyContext) {
+  const ok = await captcha(conversation, ctx);
+
+  if (ok) await ctx.reply("Welcome!");
+  else await ctx.banChatMember();
+}
+```
+
+</CodeGroupItem>
+ <CodeGroupItem title="JavaScript">
+
+```js
+async function enterGroup(conversation, ctx) {
+  const ok = await captcha(conversation, ctx);
+
+  if (ok) await ctx.reply("Welcome!");
+  else await ctx.banChatMember();
+}
+```
+
+</CodeGroupItem>
+</CodeGroup>
+
+See how the captcha function can be reused in different places in your code.
+
+> This simple example is only meant to illustrate how functions work.
+> In reality, it may work poorly because it only waits for a new update from the respective chat, but without verifying that it actually comes from the same user who joined.
+> If you want to create a real captcha, you may want to use [parallel conversations](#parallel-conversations).
+
+If you want, you can also split your code across even more fuctions, or use recursion, mutual recursion, generators, and so on.
+(Just make sure that all functions follow [the three rules](#golden-rules-of-conversations).)
+
+Naturally, you can use error handling in your functions, too.
+Regular `try`/`catch` statements work just fine, also across functions.
+After all, conversations are just JavaScript.
+
+If the main conversation function throws, the error will propagate further into [the error handling mechanisms](/guide/errors.md) of your bot.
+
+## Modules and Classes
+
+Naturally, you can just move your functions across modules.
+That way, you can define some functions in one file, `export` them, and then `import` and use them in another file.
+
+If you want, you can also define classes.
+
+<CodeGroup>
+  <CodeGroupItem title="TypeScript" active>
+
+```ts
+class Auth {
+  public token?: string;
+
+  constructor(private conversation: MyConversation) {}
+
+  authenticate(ctx: MyContext) {
+    const link = getAuthLink(); // get auth link from your system
+    await ctx.reply(
+      "Open this link to obtain a token, and sent it back to me: " + link,
+    );
+    ctx = await this.conversation.wait();
+    this.token = ctx.message?.text;
+  }
+
+  isAuthenticated(): this is Auth & { token: string } {
+    return this.token !== undefined;
+  }
+}
+
+async function askForToken(conversation: MyConversation, ctx: MyContext) {
+  const auth = new Auth(conversation);
+  await auth.authenticate(ctx);
+  if (auth.isAuthenticated()) {
+    const token = auth.token;
+    // do stuff with token
+  }
+}
+```
+
+</CodeGroupItem>
+ <CodeGroupItem title="JavaScript">
+
+```ts
+class Auth {
+  constructor(conversation) {
+    this.conversation = conversation;
```

Comment:
Maybe `this.conversation` can be `this.#conversation` as it is private that way.